src/searcher/mod.rs

```diff
@@ -297,6 +334,60 @@ impl Searcher {
         MultiLine::new(self, query_context, slice, write_to).run()
     }
 
+    fn search_slice_callback(
+        &mut self,
+        query_context: QueryContext,
+        slice: &[u8],
+        callback: impl Fn(Node, &[u8], &Path),
+        path: &Path,
+    ) -> Result<(), ConfigError> {
+        self.check_config()?;
+
+        log::trace!("slice reader: searching via multiline strategy");
+        self.run_with_callback(query_context, slice, callback, path);
+
+        Ok(())
+    }
+
+    fn run_with_callback(
+        &self,
+        query_context: QueryContext,
+        slice: &[u8],
+        callback: impl Fn(Node, &[u8], &Path),
+        path: &Path,
+    ) {
+        let mut query_cursor = QueryCursor::new();
+        let tree = get_parser(query_context.language)
+            .parse(slice, None)
+            .unwrap();
+        let query = &query_context.query;
+        let capture_index = query_context.capture_index;
+        let filter = &query_context.filter;
+        query_cursor
+            .captures(query, tree.root_node(), slice)
+            .filter_map(|(match_, found_capture_index)| {
+                let found_capture_index = found_capture_index as u32;
+                if found_capture_index != capture_index {
+                    return None;
+                }
+                let mut nodes_for_this_capture = match_.nodes_for_capture_index(capture_index);
+                let single_captured_node = nodes_for_this_capture.next().unwrap();
+                assert!(
+                    nodes_for_this_capture.next().is_none(),
+                    "I guess .captures() always wraps up the single capture like this?"
+                );
+                match filter.as_ref() {
+                    None => Some(single_captured_node),
+                    Some(filter) => filter
+                        .call(&single_captured_node)
+                        .then_some(single_captured_node),
+                }
+            })
+            .for_each(|node| {
+                callback(node, slice, path);
```

Comment:
No sinks, no line-orientation, no printer, just give the callback basically the raw match data (ie the `tree_sitter::Node`) from the tree-sitter API